examples/vector_databases/Vector_db_introduction.ipynb

```diff
@@ -0,0 +1,1187 @@
+    "from typing import List, Iterator\n",
+    "import concurrent\n",
+    "from tqdm import tqdm\n",
+    "import pandas as pd\n",
+    "from datasets import load_dataset\n",
+    "import numpy as np\n",
+    "import os\n",
+    "\n",
+    "# Pinecone's client library for Python\n",
+    "import pinecone\n",
+    "\n",
+    "# Weaviate's client library for Python\n",
+    "import weaviate\n",
+    "\n",
+    "# I've set this to our new embeddings model, this can be changed to the embedding model of your choice\n",
+    "MODEL = \"text-embedding-ada-002\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e5d9d2e1",
+   "metadata": {},
+   "source": [
+    "## Load data\n",
+    "\n",
+    "In this section we'll source the data for this task, embed it and format it for insertion into a vector database\n",
+    "\n",
+    "*Thanks to Ryan Greene for the template used for the batch ingestion"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "bd99e08e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@retry(wait=wait_random_exponential(min=1, max=20), stop=stop_after_attempt(6))\n",
+    "\n",
+    "# Simple function to take in a list of text objects and return them as a list of embeddings\n",
+    "def get_embeddings(input: List):\n",
+    "    response = openai.Embedding.create(\n",
+    "        input=input,\n",
+    "        model=MODEL,\n",
+    "    )[\"data\"]\n",
+    "    return [data[\"embedding\"] for data in response]\n",
+    "\n",
+    "# Function for batching and parallel processing the embeddings\n",
+    "def embed_corpus(\n",
+    "    corpus: List[str],\n",
+    "    batch_size=64,\n",
+    "    num_workers=8,\n",
+    "    max_context_len=8191,\n",
+    "):\n",
+    "    def batchify(iterable, n=1):\n",
+    "        l = len(iterable)\n",
+    "        for ndx in range(0, l, n):\n",
+    "            yield iterable[ndx : min(ndx + n, l)]\n",
+    "\n",
+    "    # Encode the corpus, truncating to max_context_len\n",
+    "    encoding = tiktoken.get_encoding(\"cl100k_base\")\n",
+    "    encoded_corpus = [\n",
+    "        encoded_article[:max_context_len] for encoded_article in encoding.encode_batch(corpus)\n",
+    "    ]\n",
+    "\n",
+    "    # Calculate corpus statistics: the number of inputs, the total number of tokens, and the estimated cost to embed\n",
+    "    num_tokens = sum(len(article) for article in encoded_corpus)\n",
+    "    cost_to_embed_tokens = num_tokens / 1_000 * 0.0004\n",
+    "    print(\n",
+    "        f\"num_articles={len(encoded_corpus)}, num_tokens={num_tokens}, est_embedding_cost={cost_to_embed_tokens:.2f} USD\"\n",
+    "    )\n",
+    "\n",
+    "    # Embed the corpus\n",
+    "    with concurrent.futures.ThreadPoolExecutor(max_workers=num_workers) as executor:\n",
+    "        futures = [\n",
+    "            executor.submit(get_embeddings, text_batch)\n",
```

Comment:
I've wrapped it in a try/except block so the function will return the exception if get_embeddings fails

Reply:
I mostly wasn't sure if the Exception would be raised, or if the thread would just fail silently. It seems it would be raised when `result()` is called. If that's the case you can just let the Exception propagate naturally if you think that makes the code cleaner (without the need for the try/except)

Reply:
Agree, I've removed the try/except block to make it cleaner